src/Microsoft.Extensions.OData.Migration/UriSegmentTranslator.cs

```diff
@@ -0,0 +1,243 @@
+﻿using Microsoft.OData.Edm;
+using Microsoft.OData.UriParser;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Extensions.OData.Migration
+{
+    /// <summary>
+    /// UriSegmentTranslator contains logic to translate every kind of V3 path segment into
+    /// its corresponding V4 counterpart path segment.
+    /// </summary>
+    public class UriSegmentTranslator : Data.OData.Query.SemanticAst.PathSegmentTranslator<ODataPathSegment>
+    {
+        private readonly IEdmModel v4model;
+
+        /// <summary>
+        /// Initialize UriSegmentTranslator
+        /// </summary>
+        /// <param name="v4model">v4 model is used to look up corresponding entities/etc. from v3 to create v4 segments</param>
+        public UriSegmentTranslator(IEdmModel v4model)
+        {
+            this.v4model = v4model;
+        }
+
+        /// <summary>
+        /// Translate a v3 TypeSegment (TODO: EXAMPLE) into v4 TypeSegment
+        /// </summary>
+        /// <param name="segment">OData V3 TypeSegment</param>
+        /// <returns>OData V4 TypeSegment</returns>        
+        public override ODataPathSegment Translate(Data.OData.Query.SemanticAst.TypeSegment segment)
+        {
+            // Look up v4 type by name of v3 type.
+            IEdmType v4type = v4model.FindType(segment.EdmType.FullTypeName());

```

Comment:
What if v4type is not found?

Reply:
Implemented null checking and error message across segment translator